### PR TITLE
Fix #10691 Improve Tile Inspector Delete 

### LINKED ISSUE #10691: Improve Tile Inspector Delete
When deleting tiles with the tile inspector it shouldn't be possible to delete all the tiles at one coordinate. There should always be at least one (it could also be argued there should always be a surface element as many functions assume that).

### PR DESCRIPTION
**Pull request by kylegarza21**
**When a user attempts to remove the last element on a tile, they are warned of the illegal action**
prevent_removal_of_last_element

**A user is still able to remove elements to allow for a 'floating' effect**
floating_element_allowed

### CODE REVIEW

File: src/openrct2/world/TileInspector.cpp [hunk 1]
```diff
@@ -151,12 +151,14 @@ GameActionResult::Ptr tile_inspector_remove_element_at(const CoordsXY& loc, int1
 {
     if (isExecuting)
     {
-        // Forcefully remove the element
-        TileElement* const tileElement = map_get_nth_element_at(loc, elementIndex);
-        if (tileElement == nullptr)
+        // At least one surface element must exist on a tile
+        if (1 >= windowTileInspectorElementCount)
```
Comment (AaronVanGeffen) [inline src/openrct2/world/TileInspector.cpp]: We don't use [Yoda conditions](https://en.wikipedia.org/wiki/Yoda_conditions) in this project.

```suggestion
        if (windowTileInspectorElementCount < 1)
```

File: src/openrct2/localisation/StringIds.h [hunk 2]
```diff
@@ -3905,9 +3905,10 @@ enum
 
     STR_CREATE_DUCKS_TIP = 6356,
     STR_REMOVE_DUCKS_TIP = 6357,
+    STR_TILE_NEEDS_ATLEAST_ONE_ELEMENT = 6358,
 
     // Have to include resource strings (from scenarios and objects) for the time being now that language is partially working
-    STR_COUNT = 32768
+    STR_COUNT = 32769
```
Comment (AaronVanGeffen) [inline src/openrct2/localisation/StringIds.h]: This constant should not be changed.
```suggestion
    STR_COUNT = 32768
```

**Comment by AaronVanGeffen** [comment]
Looks good to me, but I'd like another team member to review it, preferably @duncanspumpkin or @Broxzier

File: data/language/en-GB.txt [hunk 3]
```diff
@@ -3676,6 +3676,7 @@ STR_6354    :{SMALLFONT}{BLACK}Medium density
 STR_6355    :{SMALLFONT}{BLACK}High density
 STR_6356    :{SMALLFONT}{BLACK}Spawns ducks if park contains water
 STR_6357    :{SMALLFONT}{BLACK}Removes all ducks from the map
+STR_6358    :A tile must have at least one surface element
```
Comment (Broxzier) [inline data/language/en-GB.txt]:
```suggestion
STR_6358    :A tile must have at least one tile element
```

File: src/openrct2/world/TileInspector.cpp [hunk 4]
```diff
@@ -151,12 +151,15 @@ GameActionResult::Ptr tile_inspector_remove_element_at(const CoordsXY& loc, int1
 {
     if (isExecuting)
     {
-        // Forcefully remove the element
-        TileElement* const tileElement = map_get_nth_element_at(loc, elementIndex);
-        if (tileElement == nullptr)
+        // At least one surface element must exist on a tile
```
Comment (Broxzier) [inline src/openrct2/world/TileInspector.cpp]:
```suggestion
        // At least one tile element must exist on a tile
```

File: data/language/en-GB.txt [hunk 5]
```diff
@@ -3676,6 +3676,7 @@ STR_6354    :{SMALLFONT}{BLACK}Medium density
 STR_6355    :{SMALLFONT}{BLACK}High density
 STR_6356    :{SMALLFONT}{BLACK}Spawns ducks if park contains water
 STR_6357    :{SMALLFONT}{BLACK}Removes all ducks from the map
+STR_6358    :A tile must have at least one tile element
```
Comment (AaronVanGeffen) [inline data/language/en-GB.txt]: I'm not too keen on this new phrasing, as it now repeats the word 'tile'. What would be clearer? Not sure, myself.

File: src/openrct2/localisation/StringIds.h [hunk 6]
```diff
@@ -3905,6 +3905,7 @@ enum
 
     STR_CREATE_DUCKS_TIP = 6356,
     STR_REMOVE_DUCKS_TIP = 6357,
+    STR_TILE_NEEDS_ATLEAST_ONE_ELEMENT = 6358,
```
Comment (Broxzier) [inline src/openrct2/localisation/StringIds.h]: A small nitpicky thing, since this string is unrelated to the two above please add a white line this one.

File: data/language/en-GB.txt (hunk 5 above)
Comment (Broxzier) [reply data/language/en-GB.txt]: I think it's fine, "Tile" and "Tile element" are two separate things. In Dutch the latter would be one word.

Comment (tupaschoal) [reply data/language/en-GB.txt]: "A tile must have at least one element" looks fine to me, but I'm not too concerned with the repetition

File: distribution/changelog.txt [hunk 7]
```diff
@@ -27,6 +27,7 @@
 - Fix: [#10547] RCT1 parks have too many rides available.
 - Fix: [#10587] Update last action coordinates on correct player.
 - Fix: [#10662] Duck cheat tooltips look odd and do not explain anything.
+- Fix: [#10691] Improved tile inspector delete functionality.
```
Comment (AaronVanGeffen) [inline distribution/changelog.txt]: I think this should be more specific. Perhaps something like:
```suggestion
- Fix: [#10691] Improved error reporting for deleting elements in the tile inspector.
```

Comment (Broxzier) [inline distribution/changelog.txt]: This would fit better below with the other 'Improved' items, and more descriptive. Something like:
`- Improved: [#10691] Tile elements can now only be removed when there are other tile elements left.`

**Comment by Broxzier** [comment]
Ideally, the remove-button would be greyed out when it cannot be used, but that can be done later with another PR.

File: src/openrct2/world/TileInspector.cpp [hunk 8]
```diff
@@ -151,12 +151,15 @@ GameActionResult::Ptr tile_inspector_remove_element_at(const CoordsXY& loc, int1
 {
     if (isExecuting)
     {
-        // Forcefully remove the element
-        TileElement* const tileElement = map_get_nth_element_at(loc, elementIndex);
-        if (tileElement == nullptr)
+        // At least one tile element must exist on a tile
+        if (windowTileInspectorElementCount <= 1)
```
Comment (duncanspumpkin) [inline src/openrct2/world/TileInspector.cpp]: Although this does work this is not network safe. This variable is not passed over the network. You will need to recalculate the number of elements in this function to make it safe. Also this check only needs to be performed when in the non isExecuting branch.

Comment (kylegarza21) [reply src/openrct2/world/TileInspector.cpp]: @Broxzier thanks for the lead. I plan on working this, this weekend. Will let you know if there's any questions

**Comment by duncanspumpkin** [comment]
For some reason there isn't a link to this in #10691. This comment is just to show the link.

**Comment by tupaschoal** [comment]
@kylegarza21 I'm closing this due to inactivity. If you can resume the work to address the notes, feel free to reopen the PR
